# Задание 1

## Работа с библиотекой pandas

### *Цель работы:* познакомиться с использованием библиотеки pandas при решении практических задач, связанных с анализом данных.

Проанализируем датасет **employees**.  Он содержит информацию о сотрудниках организации. Набор данных включает персональные, временные и финансовые характеристики работников. Смысл задачи - на основе данных о служащих компании проанализировать структуру персонала, уровень заработной платы и бонусов, распределение по отделам и полу, а также выявить возможное гендерное неравенство. Набор данных включает в себя следующие атрибуты:

- **first name** — имя сотрудника;
- **gender** — пол сотрудника;
- **start date** — дата приёма на работу;
- **last login time** — время последнего входа в систему;
- **salary** — заработная плата сотрудника;
- **bonus %** — размер бонуса в процентах;
- **senior management** — принадлежность к высшему руководству;
- **team** — отдел, в котором работает сотрудник.

# 1. Импортируйте необходимые библиотеки

In [35]:
import pandas as pn
import numpy as np

# 2. Импортируйте набор данных из файла employees.сsv и поместите его в объект DataFrame

In [36]:
df = pn.read_csv("employees.csv")

# 3. Выведите 12 первых строк

In [37]:
print(df.head(12))

   First Name  Gender  Start Date Last Login Time  Salary  Bonus %  \
0     Douglas    Male    8/6/1993        12:42 PM   97308    6.945   
1      Thomas    Male   3/31/1996         6:53 AM   61933    4.170   
2       Maria  Female   4/23/1993        11:17 AM  130590   11.858   
3       Jerry    Male    3/4/2005         1:00 PM  138705    9.340   
4       Larry    Male   1/24/1998         4:47 PM  101004    1.389   
5      Dennis    Male   4/18/1987         1:35 AM  115163   10.125   
6        Ruby  Female   8/17/1987         4:20 PM   65476   10.012   
7         NaN  Female   7/20/2015        10:43 AM   45906   11.598   
8      Angela  Female  11/22/2005         6:29 AM   95570   18.523   
9     Frances  Female    8/8/2002         6:51 AM  139852    7.524   
10     Louise  Female   8/12/1980         9:01 AM   63241   15.132   
11      Julie  Female  10/26/1997         3:19 PM  102508   12.637   

   Senior Management                  Team  
0               True             Marketing  

# 4. Получите сводную информацию по датафрейму

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   First Name         933 non-null    object 
 1   Gender             855 non-null    object 
 2   Start Date         1000 non-null   object 
 3   Last Login Time    1000 non-null   object 
 4   Salary             1000 non-null   int64  
 5   Bonus %            1000 non-null   float64
 6   Senior Management  933 non-null    object 
 7   Team               957 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 62.6+ KB


# 5. Проверьте набор данных на наличие отсутствующих значений и дубликатов строк. При наличии обработайте их тем или иным способом.

Проверка на наличие дубликатов

In [39]:
duplicate = df[df.duplicated()]
print(duplicate)

Empty DataFrame
Columns: [First Name, Gender, Start Date, Last Login Time, Salary, Bonus %, Senior Management, Team]
Index: []


Проверим на наличие пропусков

In [40]:
df.isnull().sum()

First Name            67
Gender               145
Start Date             0
Last Login Time        0
Salary                 0
Bonus %                0
Senior Management     67
Team                  43
dtype: int64

Обработаем все данные, чтобы избавиться от пропусков

In [45]:
df = df.dropna(subset=['First Name'])

df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['Senior Management'] = df['Senior Management'].fillna(False)
df['Team'] = df['Team'].fillna('Unknown')

Проверка

In [46]:
df.isnull().sum()

First Name           0
Gender               0
Start Date           0
Last Login Time      0
Salary               0
Bonus %              0
Senior Management    0
Team                 0
dtype: int64

# 6. Получите сводку статистик для числового столбца bonus %

In [47]:
df["Bonus %"].describe()

count    933.000000
mean      10.225737
std        5.549542
min        1.015000
25%        5.359000
50%        9.849000
75%       14.879000
max       19.944000
Name: Bonus %, dtype: float64

# 7. Что является индексом этого набора данных?

In [48]:
df.index

Index([  0,   1,   2,   3,   4,   5,   6,   8,   9,  10,
       ...
       990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
      dtype='int64', length=933)

# 8. Кто из мужчин является ветераном компании?

In [49]:
df["Start Date"] = pn.to_datetime(df["Start Date"])
veteran = df[df["Gender"] == "Male"].loc[df[df["Gender"] == "Male"]["Start Date"].idxmin()]
veteran

First Name                           Joe
Gender                              Male
Start Date           1980-01-19 00:00:00
Last Login Time                  4:06 PM
Salary                            119667
Bonus %                            1.148
Senior Management                   True
Team                             Finance
Name: 954, dtype: object

# 9. Отсортируйте служащих по убыванию заработной платы и найдите пять человек, получающих наибольшую заработную плату

In [50]:
top_salary = df.sort_values("Salary", ascending=False).head(5)
print(top_salary)

    First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
644  Katherine  Female 1996-08-13        12:21 AM  149908   18.912   
429       Rose  Female 2015-05-28         8:40 AM  149903    5.630   
828    Cynthia  Female 2006-07-12         8:55 AM  149684    7.864   
160      Kathy  Female 2000-03-18         7:26 PM  149563   16.991   
740    Russell  Female 2009-05-09        11:59 AM  149456    3.533   

     Senior Management             Team  
644              False          Finance  
429              False  Human Resources  
828              False          Product  
160               True          Finance  
740              False        Marketing  


# 10.Определите наиболее распространенное в компании мужское и женское имя

In [51]:
most_common_male = df[df["Gender"] == "Male"]["First Name"].mode()[0]
most_common_female = df[df["Gender"] == "Female"]["First Name"].mode()[0]

print(most_common_male)
print(most_common_female)

Jeremy
Marilyn


# 11. Какой процент служащих относится к высшему руководству?

In [52]:
senior_pct = (df["Senior Management"].sum() / len(df)) * 100
print(senior_pct)

50.160771704180064


# 12. Определите пять служащих, входивших в систему самыми последними

In [53]:
df["Last Login Time"] = pn.to_datetime(df["Last Login Time"])
latest_logins = df.sort_values("Last Login Time", ascending=False).head(5)
print(latest_logins)

    First Name  Gender Start Date     Last Login Time  Salary  Bonus %  \
792       Anne  Female 1996-04-18 2026-04-14 23:57:00  122762    9.564   
66       Nancy  Female 2012-12-15 2026-04-14 23:57:00  125250    2.672   
930      Nancy  Female 2001-09-10 2026-04-14 23:57:00   85213    2.386   
349    Phyllis  Female 2005-11-24 2026-04-14 23:57:00  140347    8.723   
241     Walter    Male 1983-07-06 2026-04-14 23:54:00  127813    5.961   

     Senior Management                  Team  
792              False          Distribution  
66                True  Business Development  
930               True             Marketing  
349              False                 Sales  
241              False               Finance  


E:\Users\sine\AppData\Local\Temp\ipykernel_27124\1556157158.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df["Last Login Time"] = pn.to_datetime(df["Last Login Time"])


# 13.Имеет ли место гендерное неравенство в отношении заработной платы и бонусов?

In [54]:
gender_stats = df.groupby("Gender")[["Salary", "Bonus %"]].mean()
gender_stats

,Salary,Bonus %
Gender,,
Female,89958.349442,10.112487
Male,90738.873418,10.379987


Средняя заработная плата у мужчин составляет 90738.87, у женщин - 89958.35, что указывает на небольшое превышение (~780). Аналогично, средний бонус у мужчин (10.38%) незначительно выше, чем у женщин (10.11%).

Таким образом, различия присутствуют, однако они не являются существенными, поэтому нельзя однозначно говорить о выраженном гендерном неравенстве в данном наборе данных.

# 14. В каком отделе работает больше всего женщин?

In [55]:
top_women_team = df[df["Gender"] == "Female"]["Team"].mode()[0]
top_women_team

'Client Services'

# 15. Какова средняя заработная плата и средние бонусы в компании?

In [23]:
avg_salary = df["Salary"].mean() #Cредняя зп
avg_bonus = df["Bonus %"].mean() #Средний бонус

print(avg_salary)
print(avg_bonus)

90288.79635584138
10.225737406216506
